### ISSUE TITLE
UHK 80 text macro action mistypings

### ISSUE THREAD

**Issue by mondalaci** [UltimateHackingKeyboard/firmware#1000]
* Sometimes, instead of `forum.ultimatehackingkeyboard.com`, `forum.ultimaethackignkeyboard.cmo` is outputted.
* rollbar.com -> rolblar.com
* x.com/notifications -> xc.om/notifications

**Comment by mondalaci** [comment]
Removing the "Maybe fixed?" label because I can still reproduce this issue.

**Comment by pcooke9** [comment]
I've seen this a few times recently. My OS is always Win 11, and I've only noticed it while using a macro that types into the Windows Run box. It usually happens when trying to launch various apps.

I have macros to switch Windows Power Plans.
example:
<img width="539" alt="PWR Macro" src="https://github.com/user-attachments/assets/20f0eea5-66d7-40fe-abc7-e9c97cf6007a">

Today it output "powercffg -s" and "poowercfg -s", instead of "powercfg -s".

I'm using the master branch of both Agent80 and firmware-uhk80. I was connected with the spiral cable, and USB on both halves. My keystroke delay is set to 10ms.

**Comment by Zetaphor** [comment]
I use typing macros multiple times per day for both repetitive text and shortcuts. I'm still encountering this issue many times per day. It's gotten better with later firmware versions, but "better" is still VERY often, where it used to be almost unusable.

**Comment by mhantsch** [comment]
Does this problem go away if you select a keystroke delay?

![image](https://github.com/user-attachments/assets/079309fb-5426-4637-9d7a-c82a66b9f90f)

**Comment by Zetaphor** [comment]
I need to test it a lot more, (my window switcher shortcut is my most frequent and obvious pain point) but after a quick test of mashing my longer text macros repeatedly, a 5ms delay does seem to have resolved it

**Comment by pcooke9** [comment]
I always leave my keystroke delay at 10ms. If I go any lower, some apps and especially Notepad will be almost unusable.

**Comment by Zetaphor** [comment]
I don't know that I ever configured the value for my UHK60, and if I did I likely would have set it to 0

I've fortunately never had issues with it in the past. I've put in about 3 hours of testing on this setting on the UHK80 and I can say this delay was definitely the fix for me.

**Comment by pcooke9** [comment]
I had it happen with both the UHK60v1 and v2, so I just automatically set the delay to 10ms on the UHK80. It's actually pretty rare for me on the UHK80 unless I drop the delay and really intentionally spam something into Notepad. The times it happens while using the Windows Run box feels a bit different though. It only seems to rearrange a couple letters here and there. I don't see a pattern to it.

**Comment by mhantsch** [comment]
Yeah, I also had the issue already with the UHK60 with some of my macros. So I've always put that off to some weird issue of my OS that can't keep up, and defaulted to always configuring 5ms delay. Maybe the UHK80 is more prone to that issue, but I think it's not unique to the firmware-80.

I'm really unsure if there is an underlying issue that should be resolved in the firmware, and we're just working around it with the delay.

**Comment by mondalaci** [comment]
Yes, it's a standing issue.

**Comment by kareltucek** [comment]
Well, this is a mix of multiple problems.

- The macro engine produces them in correct order at least some of the time. (I.e., there are cases where I confirmed correct order on the interface between high-level firmware logic and usb subsystem, while the application produced garbled results.)

- (Some) oses / apps have problems with accepting too fast input. This problem has been present since the beginning and is effectively solved by keystrokeDelay. This is supported by two pieces of evidence: First, above observation of recorded cases where even correct macro engine sequence results in garbled output. Second, (observed in the past with a stable uhk60 firmware) the result depends on target application - with low or zero keystroke delay, some apps work flawlessly while others suffer significantly.

- uhk80 usb subsystem sometimes producing an uninterrupted stream of usb reports. (Evidence is wireshark sniffing.) Hypotesis is that this exacerbates the above "too fast input" problem. (Evidence: I havent managed to reproduce this when uhk80 produces only state change reports. It was easy to reproduce when uhk80 was producing bambilions of reports, however, in these cases I didnt check the scancode order.)

**Comment by kareltucek** [comment]
This has been resolved by Benedek's fixes, hasn't it? (@mondalaci)

**Comment by mondalaci** [comment]
According to my limited testing, it has, although the mentioned PR is incomplete. Does this issue affect the UHK 60?

**Comment by kareltucek** [comment]
> although the mentioned PR is incomplete

Ah, I have lost track.

> Does this issue affect the UHK 60?

Afaik no



I believe it has not.

**Comment by mondalaci** [comment]
I'll retest this when Benedek's PR is complete.

**Comment by pcooke9** [comment]
Thought I'd mention this just happened to me again for the first time in quite a while on the UHK80.
A macro to launch an app via the Windows Run box output "Proggorams" instead of "Programs".

•keystroke delay = 10ms
•firmware-uhk80: 11.2.0 #stuck_keys_hunt_v4
•both halves connected to the PC via USB, and spiral bridge cable connected as well.

**Comment by mondalaci** [comment]
@pcooke9 Please retest this issue with firmware 12.1.0. I can't reproduce it anymore.

**Comment by pcooke9** [comment]
> @pcooke9 Please retest this issue with firmware 12.1.0. I can't reproduce it anymore.

I haven't had it happen since my previous comment, but it always was a rare occurrence for me. I've been keeping up to date with the firmware, and I'll comment here if I see it again.

**Comment by mondalaci** [comment]
Based on my experience, this has been resolved, so I'm closing it.
